### PR TITLE
Improvements

### PR DESCRIPTION
**Pull request by nrobinaubertin**
### Overview

This PR is a follow-up of the now closed https://github.com/canonical/bind-operator/pull/54.  
It captures the improvements made in the previous PR without adding changes related to the ACLs (as those were decided to not be added to this charm during the last roadmap session).

Highlights:
- replace `ops` with `juju.application` in a lot of situations, fixing typing issues with mypy
- add a way to specify a file for the workload snap, mainly for debugging purposes.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

File: config.yaml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+options:
```
Comment (amandahla) [inline config.yaml:3]: Why a config file with no options?

Comment (nrobinaubertin) [reply config.yaml]: My bad, I removed it

File: src/bind.py [hunk 2]
```diff
@@ -330,3 +350,23 @@ def _bind_config_ip_list(self, ips: list[pydantic.IPvAnyAddress]) -> str:
...
+        Args:
+            config: dict of configuration values
+
+        Raises:
+            ConfigureError: when encountering a SnapError
+        """
+        try:
+            cache = snap.SnapCache()
+            charmed_bind = cache[constants.DNS_SNAP_NAME]
+            charmed_bind.set(config)
```
Comment (amandahla) [inline src/bind.py:366]: Does the set also reload/restart the application if needed?

File: src/charm.py [hunk 3]
```diff
@@ -147,7 +156,7 @@ def _on_config_changed(self, _: ops.ConfigChangedEvent) -> None:
     def _on_install(self, _: ops.InstallEvent) -> None:
         """Handle install."""
         self.unit.status = ops.MaintenanceStatus("Preparing bind")
```
Comment (amandahla) [inline src/charm.py:158]: If something goes wrong in the setup...What will the status of the unit be? 
Would the collect status changing it for active even if something unexpected happens?

File: src/bind.py (hunk 2 above)
Comment (nrobinaubertin) [reply src/bind.py:366]: `charmed-bind` is doing that by itself: https://github.com/canonical/charmed-bind/blob/main/snap/hooks/configure#L18

File: src/bind.py [hunk 4]
```diff
@@ -227,25 +234,38 @@ def update_zonefiles_and_reload(
...
+            # If a snap resource was not given, install the snap as published on snapcraft
+            if snap_path == "":
+                snap_cache = snap.SnapCache()
+                snap_package = snap_cache[snap_name]
+
+                if not snap_package.present or refresh:
+                    snap_package.ensure(snap.SnapState.Latest, channel=snap_channel)
+            elif pathlib.Path(snap_path).is_file():
+                # Installing the charm via subprocess.
+                # Calling subprocess here is not a security issue.
```
Comment (amandahla) [inline src/bind.py:260]: Would make sense adding a log here as information that snap is being installed with the specific path?

File: src/bind.py [hunk 5]
```diff
@@ -227,25 +234,38 @@ def update_zonefiles_and_reload(
...
+                subprocess.check_output(
+                    ["sudo", "snap", "install", snap_path, "--dangerous"]
+                )  # nosec
+            else:
+                logger.warning(
```
Comment (amandahla) [inline src/bind.py:269]: minor/nitpick/not blocker: you could return at the beginning if snap_path has value but is not a file.
I feel this could be something like "if valid snap_path  install from file else install from store".

File: src/bind.py (hunk 2 above)
Comment (amandahla) [reply src/bind.py:370]: Also, is it needed to check the present before the set?

File: src/charm.py (hunk 3 above)
Comment (nrobinaubertin) [reply src/charm.py:158]: Good question. I have to test that

File: src/bind.py [hunk 6]
```diff
@@ -44,6 +45,10 @@ class InstallError(SnapError):
     """Exception raised when unable to install dependencies for the service."""
 
 
+class ConfigureError(SnapError):
```
Comment (amandahla) [inline src/bind.py:48]: Why was ConfigureError added? Will it be handled separately in future code?

Comment (nrobinaubertin) [reply src/bind.py]: It is used by configure() but configure() itself is actually a leftover from the parent PR. I have removed both

**Review by arturo-seijas — DISMISSED** [review]

File: src/charm.py (hunk 3 above)
Comment (nrobinaubertin) [reply src/charm.py:158]: After testing I confirm that the status is updated for `hook install fail`, and the charm ends up in an error state

File: src/bind.py (hunk 5 above)
Comment (nrobinaubertin) [reply src/bind.py]: After some discussions, I decided to simplify this function by relying on the exception that would be raised by the subprocess call.

**Review by arturo-seijas — DISMISSED** [review]

File: src/bind.py [hunk 7]
```diff
@@ -227,26 +230,38 @@ def update_zonefiles_and_reload(
...
         """
         try:
-            snap_cache = snap.SnapCache()
-            snap_package = snap_cache[snap_name]
-
-            if not snap_package.present or refresh:
-                snap_package.ensure(snap.SnapState.Latest, channel=snap_channel)
-
-        except (snap.SnapError, snap.SnapNotFoundError) as e:
+            # If a snap resource was not given, install the snap as published on snapcraft
```
Comment (amandahla) [inline src/bind.py:247]: nice :)

File: actions.yaml [hunk 8]
```diff
@@ -0,0 +1,2 @@
+# Copyright 2024 Canonical Ltd.
```
Comment (amandahla) [inline actions.yaml:1]: empty actions file?

File: src/bind.py [hunk 9]
```diff
@@ -227,26 +230,38 @@ def update_zonefiles_and_reload(
...
+                logger.info(
+                    "Installing from custom charmed-bind snap located: %s",
+                    snap_path,
+                )
+                subprocess.check_output(
+                    ["sudo", "snap", "install", snap_path, "--dangerous"]
+                )  # nosec
+        except (snap.SnapError, snap.SnapNotFoundError, subprocess.CalledProcessError) as e:
             error_msg = f"An exception occurred when installing {snap_name}. Reason: {e}"
             logger.error(error_msg)
```
Comment (amandahla) [inline src/bind.py:266]: minor: logger.exception so the traceback will be logged?

Comment (nrobinaubertin) [reply src/bind.py]: Done

File: actions.yaml (hunk 8 above)
Comment (nrobinaubertin) [reply actions.yaml]: Removed

**Review by amandahla — DISMISSED** [review]

File: src/charm.py [hunk 10]
```diff
@@ -63,10 +63,19 @@ def __init__(self, *args: typing.Any):
         self.framework.observe(
             self.on[constants.PEER].relation_joined, self._on_peer_relation_joined
         )
-        self.unit.open_port("tcp", 8080)  # ACL API
         self.unit.open_port("tcp", 53)  # Bind DNS
         self.unit.open_port("udp", 53)  # Bind DNS
 
+        # Try to check if the `charmed-bind-snap` resource is defined.
```
Comment (jdkandersson) [inline src/charm.py:69]: Perhaps reword this to `Record if the ... resource is defined`. From reading `Try to check` I'm not sure what is meant by that. The code seems to set `self.snap_path` if the resource is defined

File: src/charm.py [hunk 11]
```diff
@@ -63,10 +63,19 @@ def __init__(self, *args: typing.Any):
...
+        # Using this can be useful when debugging locally
+        # More information about resources:
+        # https://juju.is/docs/sdk/resources#heading--add-a-resource-to-a-charm
+        self.snap_path: str = ""
```
Comment (jdkandersson) [inline src/charm.py:73]: I think the intent here is to note if the `snap_path` is not defined. An empty string is not always clear because an empty string could be a valid value. It is generally better to be explicit and make it `None` if there is no resource that is defined and only put a string if there is a resource to make the type clearly reflect whether the resource is set

File: tests/integration/helpers.py [hunk 12]
```diff
@@ -9,19 +9,26 @@
 # Ignore functions having too many arguments (will be removed in the future)
 # pylint: disable=too-many-positional-arguments
 
+# Ignore too many args warning. I NEED THEM OKAY ?!
+# pylint: disable=too-many-arguments
```
Comment (jdkandersson) [inline tests/integration/helpers.py:13]: I think you can also disable this for all test files in the pylint configuration, I think we reached that agreement in The Hague

File: src/charm.py (hunk 10 above)
Comment (nrobinaubertin) [reply src/charm.py]: reworded

File: src/charm.py (hunk 11 above)
Comment (nrobinaubertin) [reply src/charm.py]: I've initialized snap_path to `None`

**Review by arturo-seijas — APPROVED** [review]

**Review by javierdelapuente — APPROVED** [review]
